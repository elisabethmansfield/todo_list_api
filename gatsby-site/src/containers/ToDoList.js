import React from 'react';
// import { Location } from '@reach/router';
import SubHeader from '../components/SubHeader';
import AddTask from '../components/AddTask';
import Tasks from '../components/Tasks';
// import { Query } from 'react-apollo';
import { GET_TODOS,
         GET_TODO,
} from './queries/queries';
import{ ADD_TODO,
        DELETE_TODO,
        CHECK_TODO,
        UNCHECK_TODO,
} from './queries/mutations';
// import gql from 'graphql-tag';

export default class ToDoList extends React.Component {

  constructor(props) {
    super(props);
    this.state = {
      client: this.props.client || false,
      tasks: (this.props.data) ? this.props.data.todos : []
    };
    this.handleDeleteTask = this.handleDeleteTask.bind(this);
    this.handleAddTask = this.handleAddTask.bind(this);
    this.handleCheckbox = this.handleCheckbox.bind(this);
  }
  handleDeleteTask = (task) => {
    console.log('task parameter: ',task);
    this.setState((prevState) => ({
      tasks: prevState.tasks.filter((filtertask) => filtertask._id !== task)
    }));     
    this.state.client.mutate({
      mutation: DELETE_TODO,
      variables: { _id: task }
    })
    .then(response => {
      console.log('GraphQL delete: ',response); 
    });
  };
  handleAddTask = (task) => {
    try{
      if (!task) {
        return 'Enter valid value to add item';
      } else if (this.state.tasks.indexOf(task) > -1) {
        return 'This task already exists';
      }
      this.state.client.mutate({
        mutation: ADD_TODO,
        variables: { name: task.toString(), checked: false }
      })
      .then(response => {
        console.log('GraphQL add: ',response); 
        this.setState((prevState) => ({
          tasks: prevState.tasks.concat(response.data.addToDo)
        }));    
        console.log('this.state.tasks: ',this.state.tasks);
      });
    } catch(err){
      return 'Error: ' + err;
    }
  };
  handleCheckbox = (task) => {
    let obj = this.state.tasks.find(o => o._id === task);
    console.log('obj: ',obj);
    let mutation;
    if(obj.checked){
      mutation = UNCHECK_TODO;
    } else{
      mutation = CHECK_TODO;
    }
    this.state.client.mutate({
      mutation: mutation,
      variables: { _id: task }
    })
    .then(response => {
      console.log('GraphQL checkbox response: ',response); 
  		const newTasks = (() => {
  		  const arr = this.state.tasks;
  		  arr.forEach((o) => {
  			if(o == obj)
  				o.checked = !o.checked;
  		  });
  			return arr;		
  		})();
  		console.log('newTasks: ',newTasks);
  		this.setState((prevState) => ({
  		  tasks: newTasks
  		}));
    });
  };
  componentDidMount() {
    if(this.state.client){
      this.state.client.query({
        query: GET_TODOS
      })
      .then(response => {
        console.log('GraphQL query: ',response.data.todos); 
        this.setState((prevState) => ({
           tasks: response.data.todos
        }));
        console.log('this.state.tasks: ',this.state.tasks);
      });
    }
    else{
      console.log('Apollo client is not connected.');
    }
  }
  render() {
    const subtitle = this.props.subtitle || 'Add and remove tasks';
    return (
        <div>
          <SubHeader subtitle={subtitle} />
          <div className="container">
            <div className="widget">
              {/* <Location>
                {({ location }) => {
                  console.log(location)
                  return <p>The location is {location.pathname}</p>
                }}
              </Location> */}
              <Tasks
                tasks={this.state.tasks}
                handleDeleteTask={this.handleDeleteTask}
                handleCheckbox={this.handleCheckbox}
              />
              <AddTask
                handleAddTask={this.handleAddTask}
              />
            </div>
          </div>
        </div>
    );
  }
}

  // handleDeleteTasks = () => {
  //   this.setState(() => ({ tasks: [] }));
  // };
  
// apolloClient.mutate({
//   variables: { text: "hello" },
//   mutation: gql`
//     mutation AddComment($text: String!){
//       addComment(text: $text) {
//         id
//         text
//       }
//     }
//   `,
// }) 
// handleDeleteTasks={this.handleDeleteTasks}
