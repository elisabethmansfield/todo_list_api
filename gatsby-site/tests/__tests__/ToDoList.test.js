/* https://www.gatsbyjs.org/docs/testing-components-with-graphql/
Testing components with GraphQL
*/

import React from "react";
import renderer from "react-test-renderer";
import ToDoList from "../../src/containers/ToDoList";

describe("ToDoList", () =>
  it("renders correctly", () => {

    // mock GraphQL data
    const data = {
    "todos": [
      {
        "_id": "5bf4e5e4232d4b185e963d4d",
        "name": "Hello World",
        "checked": false
      },
      {
        "_id": "6bf4e554e3dd4b385e9e3a4c",
        "name": "Go to Store",
        "checked": true
      },
      {
        "_id": "8af4e5e5432e3b181e9e3d3b",
        "name": "Buy Milk",
        "checked": false
      }
    ]};

    const tree = renderer
      .create(<ToDoList data={data} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
}));