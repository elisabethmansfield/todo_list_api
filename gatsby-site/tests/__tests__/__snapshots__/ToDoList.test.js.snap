// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToDoList renders correctly 1`] = `
<div>
  <div
    className="header"
  >
    <div
      className="container"
    >
      <h1
        className="header__title"
      >
        To Do List
      </h1>
      <h2
        className="header__subtitle"
      >
        Add and remove tasks
      </h2>
    </div>
  </div>
  <div
    className="container"
  >
    <div
      className="widget"
    >
      <div>
        <div
          className="widget-header"
        >
          <h3
            className="widget-header__title"
          >
            Task List
          </h3>
        </div>
        <div
          className="task"
          id="5bf4e5e4232d4b185e963d4d"
        >
          <div
            class="custom-control custom-checkbox"
          >
            <input
              checked={false}
              class="custom-control-input"
              id={1}
              name={1}
              onChange={[Function]}
              type="checkbox"
            />
            <label
              class="task__text"
              for={1}
            >
              Hello World
            </label>
          </div>
          <button
            className="button button--link"
            onClick={[Function]}
          >
            Remove
          </button>
        </div>
        <div
          className="task"
          id="6bf4e554e3dd4b385e9e3a4c"
        >
          <div
            class="custom-control custom-checkbox"
          >
            <input
              checked={true}
              class="custom-control-input"
              id={2}
              name={2}
              onChange={[Function]}
              type="checkbox"
            />
            <label
              class="task__text"
              for={2}
            >
              Go to Store
            </label>
          </div>
          <button
            className="button button--link"
            onClick={[Function]}
          >
            Remove
          </button>
        </div>
        <div
          className="task"
          id="8af4e5e5432e3b181e9e3d3b"
        >
          <div
            class="custom-control custom-checkbox"
          >
            <input
              checked={false}
              class="custom-control-input"
              id={3}
              name={3}
              onChange={[Function]}
              type="checkbox"
            />
            <label
              class="task__text"
              for={3}
            >
              Buy Milk
            </label>
          </div>
          <button
            className="button button--link"
            onClick={[Function]}
          >
            Remove
          </button>
        </div>
      </div>
      <div>
        <form
          className="add-task"
          onSubmit={[Function]}
        >
          <input
            className="add-task__input"
            name="task"
            type="text"
          />
          <button
            className="button"
          >
            Add Task
          </button>
        </form>
      </div>
    </div>
  </div>
</div>
`;
